src/Microsoft.Diagnostics.Monitoring/DiagnosticServices.cs

```diff
@@ -68,8 +68,25 @@ public async Task<Stream> GetDump(int pid, DumpType mode)
             return new AutoDeleteFileStream(dumpFilePath);
         }
 
-        public async Task<IStreamWithCleanup> StartCpuTrace(int pid, TimeSpan duration, CancellationToken cancellationToken)
+        public async Task<Stream> GetGcDump(int pid, TimeSpan timeout, CancellationToken cancellationToken)
         {
+            using CancellationTokenSource linkedSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            linkedSource.CancelAfter(timeout);
```

Comment:
I'll push the CancellationTokenSource creation to the DiagController and have it link in the RequestAborted token and then pass in its token to this method and remove the timeout parameter.

Reply:
I'm also questioning the timeout parameter of the REST call : ) I'm not sure why the client would need to specify a value when it could disconnect after that amount of time had passed.

Reply:
If a user is only using a browser to to interact with dotnet-monitor, then I think the only way to cancel it (if there is no timeout parameter) is to close the browser tab or the entire browser.

Reply:
Along that line of reasoning, you could almost argue that the other operations don't need duration parameters either, since the client could always just disconnect after a certain amount of time. However, using the browser as the client, that would require the user to close the browser or browser tab, and the user would have to manually save as much of the streamed output as the user wanted before closing the connection.

Reply:
I think we should follow the same pattern here as with the other commands:
- Have a reasonable default timeout.
- Allow for an infinite timeout/duration for situations where the user deems it appropriate.

Reply:
The duration parameter on some of the other controller methods has a different meaning - it tells the runtime (via the monitor) that it should gracefully stop the trace after that period of time. All trace data still gets flushed after the duration time period elapses and the resulting trace is usable and complete for the period of time requested. Duration also doesn't create a hard upper bound for the amount of time the client is willing to wait because rundown events which occur as part of the graceful end of a trace can be quite substantial for some apps causing them to transmit for seconds (or minutes even) after the stop request was received. It is only in the gc-dump case that the timeout indicates an abort, the trace is cut off abruptly and the resulting gcdump is not usable.

If the user is using a browser to manually invoke REST apis, having them close that browser tab when they don't want to wait any longer seems reasonable to me. This is what the user would be required to do for all the other REST APIs on the controller because the duration times do not enforce a hard time limit on the total amount of time spent receiving data.

Reply:
While I agree with your distinction between timeout and duration, my concern here is from a usability perspective. For example, the dotnet-gcdump tool defaults to 30 seconds.  It's a better experience if we fail after a certain timeout than if we just sit there in a non-responsive state forcing the user to manually cancel.

Reply:
> my concern here is from a usability perspective

I think we are overindexing on the scenario where users are using a browser to access the low level REST API. My hope is that we are going to offer some html UI for people with browsers and people that want to poke the raw data-collection endpoints will use tools like cURL. In curl there is a --max-time option and wget has a --timeout option. dotnet-gcdump does have a timeout as scripting ctrl-C would be pretty annoying but for a user in a browser closing the tab or hitting the stop button both seem like simple/normal actions to me? The consequence of adding default timeouts for browser users is that they are going to show up for the command-line tool users as well. A user using curl <endpoint> --max-time 60 is going to get a failure after 30 seconds because they didn't realize they need to override the server-side timeout with a query parameter.

If we do decide server-side default timeouts are important, that suggests we should be adding it to all the APIs, not only gc-dump. The duration option isn't a substitute for timeout in this sense. I could write an app where a 30 sec duration cpu trace will still be emitting events after a minute. Certainly I'd be provoking corner-case behavior to achieve that, but I think the timeout for gc-dump is also aimed at obscure cases. In the normal case all these things finish promptly and no timeout is necessary.

Reply:
Noah and I discussed this a bit offline. I think we'll need a bigger discussion around which client we want to prioritize. I think for now we can remove the timeout.